### PR TITLE
fix: validate host prefix format to prevent '%' in printf strings

### PR DESCRIPTION
**Pull request by stelfrag**
##### Summary
- Harden NETDATA_HOST_PREFIX validation to block unsafe format specifiers before they reach path templates.

---
## Summary by cubic
Blocks '%' in host prefix validation to prevent printf format string injection in path templates. Also rejects unsafe `NETDATA_HOST_PREFIX` values from the environment and adds targeted tests.

- **Bug Fixes**
  - Validate `netdata_configured_host_prefix` against '%' (check original string, not truncated); on failure, clear the prefix and set errno to `EINVAL`.
  - Guard env var assignment in `nd_log_initialize_for_external_plugins()` to reject '%' before format strings.

- **New Tests**
  - Added `paths_unittest()` with cases for embedded/escaped '%' and truncation edge cases; wired into `netdata_main()` test runner.

<sup>Written for commit d7d8eff1bbd84c707c19b240a54531ac89958496. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/netdata/netdata/pull/23342?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

**Comment by stelfrag** [comment]
@cubic-dev-ai review PR

File: src/libnetdata/paths/paths.c [hunk 1]
```diff
@@ -67,6 +67,17 @@ int verify_netdata_host_prefix(bool log_msg) {
 
     strncpyz(path, netdata_configured_host_prefix, sizeof(path) - 1);
 
+    // the host prefix is concatenated into printf format strings (the various
+    // "path to ..." templates), so a '%' in it would be parsed as a conversion
+    // directive. a real host prefix never contains one.
+    // check the original, not the truncated copy in path[] - on success the
+    // untruncated value is what stays in netdata_configured_host_prefix.
+    if(strchr(netdata_configured_host_prefix, '%')) {
```
[the remark this answers is not in the document]
Comment (stelfrag) [reply src/libnetdata/paths/paths.c]: The premise does hold — the enumerated sites are the ones that pass the prefix as a `%s` argument, but there's a second pattern in the codebase: the prefix is prepended to path *templates* that are then used **as** the format string.

`proc_diskstats.c:1370` builds `path_to_sys_block_device = "<prefix>/sys/block/%s"`, and `:781` calls:

```c
snprintfz(buffer, FILENAME_MAX, path_to_sys_block_device, disk);
```

Same shape in three more files — 17 templates in total:

| Template | Consumed as format string at |
|---|---|
| `proc_diskstats.c:1370-1379` — `path_to_sys_block_device`, `_bcache`, `path_to_sys_devices_virtual_block_device`, `path_to_sys_dev_block_major_minor_string` | `:762`, `:781`, `:789`, `:796`, `:803`, `:845` |
| `proc_stat.c:551-576` — `core_throttle_count_filename`, `package_throttle_count_filename`, `scaling_cur_freq_filename`, `time_in_state_filename`, `cpuidle_name_filename`, `cpuidle_time_filename` | `:404`, `:410`, `:417`, `:438`, `:633`, `:642`, `:651`, `:658`, `:888`, `:896` |
| `proc_net_dev.c:897-912` — `path_to_sys_devices_virtual_net`, `path_to_sys_class_net_speed`, `_duplex`, `_operstate`, `_carrier`, `_mtu` | `:986`, `:1004`, `:1011`, … |
| `proc_mdstat.c:120` — `mismatch_cnt_filename` | `:325` |

So with prefix `/host%s`, the first template becomes `"/host%s/sys/block/%s"` and is invoked with a single argument — the trailing `%s` dereferences a vararg that was never passed. `%n` in a prefix would be a write primitive. That is exactly what the `"trailing %s"` and `"bare %n"` unit-test cases cover.

On the `100%` mount being a regression: it does not work today either. `"/mnt/100%" + "/sys/block/%s"` produces `%/`, an undefined conversion specifier — glibc and musl do not treat it alike, and we build against both. So the trade is a logged rejection versus unspecified behavior, not a working configuration being broken.

Your last point was right, though, and is the real defect here: the comment said "the various 'path to ...' templates" without naming anything, which is precisely why the check read as baseless. I've updated it to record the concrete call sites above (landing in the next push).

Supporting a `%` in the prefix properly would mean escaping `%` → `%%` at all 17 template sites rather than relaxing this check — noted as such in the comment. Happy to open a separate issue for that if the `100%` case is worth supporting.

File: src/libnetdata/paths/paths.c [hunk 2]
```diff
@@ -281,3 +292,90 @@ void recursive_config_double_dir_load(const char *user_path, const char *stock_p
...
+    const struct {
+        const char *name;
+        const char *prefix;
+        bool accept;
+    } cases[] = {
+        { "empty",            "",           true  },
+#if !defined(OS_WINDOWS)
+        // needs a real procfs+sysfs under the prefix; on macOS/FreeBSD those
+        // checks are compiled out, but native Windows would probe and fail
+        { "root",             "/",          true  },
```
[the remark this answers is not in the document]
Comment (stelfrag) [reply src/libnetdata/paths/paths.c]: Valid — fixed in 1872a1e788.

Root cause: `verify_netdata_host_prefix()` only accepts a non-empty prefix after `stat()` + `is_procfs("<prefix>/proc")` + `is_sysfs("<prefix>/sys")` all succeed, so the `"root"` case was asserting host mount state rather than `%` validation. The `#if !defined(OS_WINDOWS)` guard was a compile-time approximation of a runtime condition.

The case is now skipped when the mounts are unavailable, probed with the same predicates the function itself uses so the skip condition cannot drift from the accept condition:

```c
bool host_mounts_available = is_procfs("/proc", NULL) == 0 && is_sysfs("/sys", NULL) == 0;
```

The platform guard is gone — the runtime probe covers Windows too — and the case reports `SKIPPED` instead of counting as an error. The five `%`-rejection cases and their `EINVAL` assertions are host-independent and always run, so coverage of the actual validation is unchanged.

File: src/libnetdata/log/nd_log-init.c [hunk 3]
```diff
@@ -96,7 +96,9 @@ void nd_log_initialize_for_external_plugins(const char *name) {
 
     if(!netdata_configured_host_prefix) {
         s = getenv("NETDATA_HOST_PREFIX");
-        if(s && *s)
+        // this assignment bypasses verify_netdata_host_prefix(), so apply the
+        // same rule here - see netdata_host_prefix_has_format_specifier().
+        if(s && *s && !netdata_host_prefix_has_format_specifier(s))
             netdata_configured_host_prefix = (char *)s;
     }
```
[the remark this answers is not in the document]
Comment (stelfrag) [reply src/libnetdata/log/nd_log-init.c:103]: The declaration is in scope — it arrives transitively, which is the convention in this tree.

`netdata_host_prefix_has_format_specifier()` is declared at `src/libnetdata/paths/paths.h:31`, and this file reaches it via:

```
nd_log-init.c:3            #include "nd_log-internals.h"
nd_log-internals.h:6       #include "../libnetdata.h"
libnetdata.h:29            #include "paths/paths.h"
```

That `libnetdata.h:29` include is unconditional — not inside any `#if`. Per the repository's C conventions, `libnetdata.h` includes everything in libnetdata, so libnetdata sources deliberately do not add per-header includes; adding `#include "../paths/paths.h"` here would be redundant and inconsistent with every other libnetdata TU.

On warnings-as-errors: that is exactly why this would not have survived CI. All 136 checks on this PR pass, including 52 `Build` jobs and the full `Test Source Build` matrix (glibc and musl, gcc and clang). An implicit function declaration would have failed all of them, not none.

File: src/daemon/main.c [hunk 4]
```diff
@@ -520,6 +520,7 @@ int netdata_main(int argc, char **argv) {
                             if (unittest_waiting_queue()) return 1;
                             if (rw_spinlock_unittest()) return 1;
                             if (uuidmap_unittest()) return 1;
+                            if (paths_unittest()) return 1;
 #ifdef HAVE_LIBBACKTRACE
```
[the remark this answers is not in the document]
Comment (stelfrag) [reply src/daemon/main.c:524]: Same as the `nd_log-init.c` comment — the prototype is in scope transitively.

`paths_unittest()` is declared at `src/libnetdata/paths/paths.h:33`, alongside the other `paths` declarations, and `main.c` reaches it via:

```
main.c:3                   #include "common.h"
common.h:10                #include "libnetdata/libnetdata.h"
libnetdata.h:29            #include "paths/paths.h"
```

Declaring it in `src/daemon/unit_test.h` instead would be the wrong home: `unit_test.h` holds the daemon's own unit tests, while the neighbouring calls in this same block — `rw_spinlock_unittest()`, `uuidmap_unittest()`, `json_c_parser_unittest()` — are all libnetdata tests declared next to the code they exercise. `paths_unittest()` follows that existing pattern.

CI confirms it compiles: 136/136 checks pass across the `Build` and `Test Source Build` matrices.

**Review by thiagoftsm — APPROVED** [review]
PR is working as expected. LGTM!
